encoding/ccf/decode_type.go

```diff
@@ -0,0 +1,615 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright 2022-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ccf
+
+import (
+	"fmt"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/sema"
+)
+
+type cadenceTypeID string
+
+type decodeTypeFn func(types cadenceTypeByCCFTypeID) (cadence.Type, error)
+
+// decodeInlineType decodes inline-type as
+// language=CDDL
+// inline-type =
+//
+//	simple-type
+//	/ optional-type
+//	/ varsized-array-type
+//	/ constsized-array-type
+//	/ dict-type
+//	/ reference-type
+//	/ restricted-type
+//	/ capability-type
+//	/ type-ref
+//
+// All exported Cadence types needs to be handled in this function,
+// including abstract and interface types.
+func (d *Decoder) decodeInlineType(types cadenceTypeByCCFTypeID) (cadence.Type, error) {
+	tagNum, err := d.dec.DecodeTagNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	switch tagNum {
+	case CBORTagSimpleType:
+		return d.decodeSimpleTypeID()
+
+	case CBORTagOptionalType:
+		return d.decodeOptionalType(types, d.decodeInlineType)
+
+	case CBORTagVarsizedArrayType:
+		return d.decodeVarSizedArrayType(types, d.decodeInlineType)
+
+	case CBORTagConstsizedArrayType:
+		return d.decodeConstantSizedArrayType(types, d.decodeInlineType)
+
+	case CBORTagDictType:
+		return d.decodeDictType(types, d.decodeInlineType)
+
+	case CBORTagReferenceType:
+		return d.decodeReferenceType(types, d.decodeInlineType)
+
+	case CBORTagRestrictedType:
+		return d.decodeRestrictedType(types, d.decodeInlineType)
+
+	case CBORTagCapabilityType:
+		return d.decodeCapabilityType(types, d.decodeInlineType)
+
+	case CBORTagTypeRef:
+		return d.decodeTypeRef(types)
+
+	default:
+		return nil, fmt.Errorf("unsupported encoded inline type with CBOR tag number %d", tagNum)
```

Comment:
Maybe use `errors.NewDefaultUserError()` for all decoding errors. I think it's safe to categorize any invalid input as a user error (from the decoder's POV).

Reply:
Yes, I agree.

We don't need to do it here because `Decoder.Decode()` already wraps returned error inside `DefaultUserError` like this:

https://github.com/onflow/cadence/blob/266fb6a280996e376da1e56ac228bd006ee5c2b0/encoding/ccf/decode.go#L108-L110